lib/navigation/preprocess_encoders.cpp

```diff
@@ -1,56 +1,106 @@
+#include "consts.hpp"
 #include "preprocess_encoders.hpp"
 
+#include "core/types.hpp"
+
 namespace hyped::navigation {
 
-EncodersPreprocessor::EncodersPreprocessor()
+EncodersPreprocessor::EncodersPreprocessor(core::ILogger &logger)
+    : logger_(logger),
+      num_consecutive_outliers_per_encoder_({0, 0, 0, 0}),
+      are_encoders_reliable_({true, true, true, true}),
+      num_reliable_encoders_(core::kNumEncoders),
+      max_consecutive_outliers_(10)
+{
+}
+
+std::optional<core::EncoderData> EncodersPreprocessor::processData(
+  const core::EncoderData &raw_encoder_data)
 {
-  // TODOLater: implement
+  const auto encoder_data = sanitise(raw_encoder_data);
+  const auto sensors      = checkReliable();
+  if (sensors == SensorChecks::kUnacceptable) { return std::nullopt; }
```

Comment:
I guess the checkReliable() function takes care of that.

Reply:
i think it would log when more than one encoder has failed, but I'd think we want to know even when a single encoder has failed somewhere (it looks like you could log this right after you decrement the `num_reliable_encoders` actually)

Reply:
Yeah don't be afraid of using the log. Probably use the `kWarn` level for if just one sensor becomes unreliable, `kCritical` for any fail states and `kDebug` otherwise I guess?

Reply:
`kWarn` coming from #91 soon™

Reply:
So once #91 is merged into main, merge main into this branch @aryansharma19.